File: core-render-ohos/src/main/cpp/libohos_render/expand/modules/file/KRFileModule.cpp
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "KRFileModule.h"

#include <fcntl.h>
#include <sys/stat.h>
#include <cstdio>
#include <cstring>
#include <string>
#include <thread>

#include "libohos_render/utils/KRJSONObject.h"

namespace kuikly {
namespace module {

const char KRFileModule::MODULE_NAME[]       = "KRFileModule";
const char KRFileModule::METHOD_WRITE_FILE[] = "writeFile";
const char KRFileModule::METHOD_APPEND_FILE[]  = "appendFile";
const char KRFileModule::METHOD_GET_FILES_DIR[] = "getFilesDir";

// ---------------------------------------------------------------------------
// 工具：确保目录存在
// ---------------------------------------------------------------------------
static void MkdirIfNeeded(const std::string &dir) {
    struct stat st{};
    if (stat(dir.c_str(), &st) != 0) {
        mkdir(dir.c_str(), 0755);
    }
}

// ---------------------------------------------------------------------------
// 工具：构建回调结果 map
// ---------------------------------------------------------------------------
static KRAnyValue MakeResult(const std::string &key, const std::string &value) {
    KRRenderValueMap result;
    result[key] = KRRenderValue::Make(value);
    return KRRenderValue::Make(result);
}

// ---------------------------------------------------------------------------
// 获取 profiler 写入目录（filesDir/KuiklyProfiler/）
// ---------------------------------------------------------------------------
std::string KRFileModule::GetProfilerDir() {
    auto root = GetRootView().lock();
    if (!root) {
        return "";
    }
    const std::string filesDir = root->GetContext()->Config()->GetFilesDir();
    if (filesDir.empty()) {
        return "";
    }
    std::string profilerDir = filesDir + "/KuiklyProfiler";
    MkdirIfNeeded(profilerDir);
    return profilerDir;
}

// ---------------------------------------------------------------------------
// writeFile：覆盖写，在后台线程执行
// ---------------------------------------------------------------------------
void KRFileModule::WriteFile(const KRAnyValue &params, const KRRenderCallback &callback) {
    auto jsonObj = util::JSONObject::Parse(params->toString());
    const std::string filename = jsonObj->GetString("filename");
    const std::string content  = jsonObj->GetString("content");

    if (filename.empty() || content.empty()) {
        if (callback) callback(MakeResult("error", "missing filename or content"));
        return;
    }

    const std::string dir = GetProfilerDir();
    if (dir.empty()) {
        if (callback) callback(MakeResult("error", "context unavailable"));
        return;
    }

    const std::string filePath = dir + "/" + filename;

    std::thread([filePath, content, callback]() {
        FILE *fp = fopen(filePath.c_str(), "w");
        if (!fp) {
            if (callback) callback(MakeResult("error", "fopen failed"));
            return;
        }
        fwrite(content.c_str(), 1, content.size(), fp);
        fclose(fp);
        if (callback) callback(MakeResult("path", filePath));
    }).detach();
}

// ---------------------------------------------------------------------------
// appendFile：追加写，末尾加换行（适合 JSONL），在后台线程执行
// ---------------------------------------------------------------------------
void KRFileModule::AppendFile(const KRAnyValue &params, const KRRenderCallback &callback) {
    auto jsonObj = util::JSONObject::Parse(params->toString());
    const std::string filename = jsonObj->GetString("filename");
    const std::string content  = jsonObj->GetString("content");

    if (filename.empty() || content.empty()) {
        if (callback) callback(MakeResult("error", "missing filename or content"));
        return;
    }

    const std::string dir = GetProfilerDir();
    if (dir.empty()) {
        if (callback) callback(MakeResult("error", "context unavailable"));
        return;
    }

    const std::string filePath = dir + "/" + filename;

    std::thread([filePath, content, callback]() {
        FILE *fp = fopen(filePath.c_str(), "a");
        if (!fp) {
            if (callback) callback(MakeResult("error", "fopen failed"));
            return;
        }
        // 追加内容 + 换行，适合 JSONL 格式
        fwrite(content.c_str(), 1, content.size(), fp);
        fwrite("\n", 1, 1, fp);
        fclose(fp);
        if (callback) callback(MakeResult("path", filePath));
    }).detach();
}

// ---------------------------------------------------------------------------
// getFilesDir：返回 profiler 目录路径
// ---------------------------------------------------------------------------
void KRFileModule::GetFilesDir(const KRRenderCallback &callback) {
    if (!callback) return;
    const std::string dir = GetProfilerDir();
    if (dir.empty()) {
        callback(MakeResult("error", "context unavailable"));
    } else {
        callback(MakeResult("path", dir));
    }
}

// ---------------------------------------------------------------------------
// CallMethod 分发
// ---------------------------------------------------------------------------
KRAnyValue KRFileModule::CallMethod(bool sync, const std::string &method,
                                    KRAnyValue params,
                                    const KRRenderCallback &callback) {
    if (method == METHOD_WRITE_FILE) {
        WriteFile(params, callback);
    } else if (method == METHOD_APPEND_FILE) {
        AppendFile(params, callback);
    } else if (method == METHOD_GET_FILES_DIR) {
        GetFilesDir(callback);
    }
    return KREmptyValue();
}

void KRFileModule::OnDestroy() {}

}  // namespace module
}  // namespace kuikly

File: core-render-android/src/main/java/com/tencent/kuikly/core/render/android/expand/module/KRFileModule.kt
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.kuikly.core.render.android.expand.module

import com.tencent.kuikly.core.render.android.css.ktx.toJSONObjectSafely
import com.tencent.kuikly.core.render.android.export.KuiklyRenderBaseModule
import com.tencent.kuikly.core.render.android.export.KuiklyRenderCallback
import java.io.File

/**
 * 文件读写 Module，提供 App Caches 目录的文件写入能力。
 * 主要用于 RecompositionProfiler 导出 JSON 报告供 AI 分析。
 *
 * 写入目录：context.cacheDir/KuiklyProfiler/
 * 使用 cacheDir（而非 filesDir）的原因：
 *   - 不纳入 备份
 *   - 多页面共享同一目录，同名文件后写覆盖前写
 *   - adb shell run-as <pkg> cat cache/KuiklyProfiler/<filename> 可直接读取
 */
class KRFileModule : KuiklyRenderBaseModule() {

    override fun call(method: String, params: String?, callback: KuiklyRenderCallback?): Any? {
        return when (method) {
            METHOD_WRITE_FILE -> writeFile(params, callback)
            METHOD_APPEND_FILE -> appendFile(params, callback)
            METHOD_GET_FILES_DIR -> getFilesDir(callback)
            else -> super.call(method, params, callback)
        }
    }

    private fun profilerDir(): File? {
        val cacheDir = context?.cacheDir ?: return null
        val dir = File(cacheDir, "KuiklyProfiler")
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    private fun getFilesDir(callback: KuiklyRenderCallback?) {
        val dir = profilerDir()
        if (dir != null) {
            callback?.invoke(mapOf("path" to dir.absolutePath))
        } else {
            callback?.invoke(mapOf("error" to "context unavailable"))
        }
    }

    private fun appendFile(params: String?, callback: KuiklyRenderCallback?) {
        val json = params.toJSONObjectSafely()
        val filename = json.optString(PARAM_FILENAME)
        val content = json.optString(PARAM_CONTENT)

        if (filename.isNullOrEmpty() || content.isNullOrEmpty()) {
            callback?.invoke(mapOf("error" to "missing filename or content"))
            return
        }

        val dir = profilerDir()
        if (dir == null) {
            callback?.invoke(mapOf("error" to "context unavailable"))
            return
        }

        Thread {
            try {
                val file = File(dir, filename)
                // 追加写，末尾加换行，适合 JSONL 格式
                file.appendText(content + "\n", Charsets.UTF_8)
                callback?.invoke(mapOf("path" to file.absolutePath))
            } catch (e: Exception) {
                callback?.invoke(mapOf("error" to (e.message ?: "unknown error")))
            }
        }.start()
    }

    private fun writeFile(params: String?, callback: KuiklyRenderCallback?) {
        val json = params.toJSONObjectSafely()
        val filename = json.optString(PARAM_FILENAME)
        val content = json.optString(PARAM_CONTENT)

        if (filename.isNullOrEmpty() || content.isNullOrEmpty()) {
            callback?.invoke(mapOf("error" to "missing filename or content"))
            return
        }

        val dir = profilerDir()
        if (dir == null) {
            callback?.invoke(mapOf("error" to "context unavailable"))
            return
        }

        // 在后台线程执行文件写入，避免阻塞 UI
        Thread {
            try {
                val file = File(dir, filename)
                file.writeText(content, Charsets.UTF_8)
                callback?.invoke(mapOf("path" to file.absolutePath))
            } catch (e: Exception) {
                callback?.invoke(mapOf("error" to (e.message ?: "unknown error")))
            }
        }.start()
    }

    companion object {
        const val MODULE_NAME = "KRFileModule"
        private const val METHOD_WRITE_FILE = "writeFile"
        private const val METHOD_APPEND_FILE = "appendFile"
        private const val METHOD_GET_FILES_DIR = "getFilesDir"
        private const val PARAM_FILENAME = "filename"
        private const val PARAM_CONTENT = "content"
    }
}
